### ISSUE TITLE
Level 3 of Lab 07

### ISSUE THREAD

**Issue by isaactyj** [nus-cs2030/2021-s2#156]
# Summary

I am trying to get the map function to work :/
## Description:
What I did was to have <R> Logger<R> map(Function<? super T, ? extends R> mapper)  in my interface and to code the thing out in my loggerimpl by returning a make(mapper(this.obj)). However I got a attempt to assign weaker assess to weaker assess  privilege. I had all my method to be public so I'm not sure what is happening. Can someone please advise:C

## Code Snippets

Attach code snippets (if any)

## Screenshots

Screenshots (if any)

**Comment by jethrokyq** [comment]
Hi Isaac, actually you are mapping the value within the logger to another value, so you should do funct.apply("object's value") instead. Beyond that, you need some way to pass information about the history of the previous logger to the new logger which is created. Hope this helps!

**Comment by Zekeseoj** [comment]
Hi can i also ask if it is possible for an arraylist to act as a "storage" to contain either integers or strings as shown in the test cases ? How can i do that?

**Comment by isaactyj** [comment]
@jethrokyq sorry could you explain funct.apply?

**Comment by ivankqw** [comment]
> @jethrokyq sorry could you explain funct.apply?

.apply is a method defined in the Function functional interface. Check it out here: https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/util/function/Function.html#apply(T)
 
say in map(Function<? super T, ? extends R> mapper), mapper.apply(object's value) operates on a type ? super T and returns a type ? extends R, depending on how you will define Function mapper. hope it helps

**Comment by isaactyj** [comment]
@ivankqw Cheers! for this level 3, I tried to overload the constructor to take in an old obj so as to do a comparison but in doing so I get incompatible type, how would you advise on solving this?

**Comment by jethrokyq** [comment]
@Zekeseoj this is just a suggestion but you can use some sort of data structure like a deque.
Essentially during the first make call, you will always store "Value initialized..." as the first log entry, then during subsequent map calls, since you are creating a new logger, you can overload make to take in both the new value as well as well as the updated data structure containing the logs.

**Comment by ivankqw** [comment]
> @ivankqw Cheers! for this level 3, I tried to overload the constructor to take in an old obj so as to do a comparison but in doing so I get incompatible type, how would you advise on solving this?

Hi, do elaborate more on this

**Comment by Keithlimhs** [comment]
Hi, you can overload the LoggerImpl constructor which takes in the new value after the map and a new String which is the new printlog output. Inside the map method, you can compare the mapped value with the old value and add the String according, thereafter calling the overloaded constructor with the new value and updated String

**Comment by isaactyj** [comment]
@ivankqw I tried to make it as 
    public <? extends R> Logger<? extends R> map(Function<? super T, ? extends R> mapper) {
        printlog();
        return make(mapper.apply(this.obj), this.obj);
    }
then have 2 constructors, one that takes in one obj, one that takes in 2 obj, then it will have a string log that will depend on whether the 2 obj are equal but somehow there is an error

**Comment by isaactyj** [comment]
@Keithlimhs I tried but I still keep getting incompatible type...

**Comment by Keithlimhs** [comment]
> @Keithlimhs I tried but I still keep getting incompatible type...

Your second constructor can be LoggerImpl(T value, String log), and the log can be assigned to be the output for printlog(). Then in the map method, you change the String log according to the comparison and return a new LoggerImpl object using the constructor

**Comment by isaactyj** [comment]
@Keithlimhs I tried your method but somehow I am unable to compile it.
error: incompatible types: String cannot be converted to java.lang.String
        return new LoggerImpl<T>(obj, log);

This is what I got

**Comment by Keithlimhs** [comment]
> @Keithlimhs I tried your method but somehow I am unable to compile it.
> error: incompatible types: String cannot be converted to java.lang.String
> return new LoggerImpl(obj, log);
> 
> This is what I got

hmm can i see how you implemented it

**Comment by isaactyj** [comment]
public <R> Logger<R> map(Function<? super T, ? extends R> mapper) {
        printlog();
        if(mapper.apply(this.obj) == this.obj) {
            String prompt = String.format("Value unchanged. Value =  %s", this.obj);
            return make(mapper.apply(this.obj), prompt);
        }
        String prompt = String.format("Value changed! New Value = %s", mapper.apply(this.obj));
        return make(mapper.apply(this.obj), prompt);
    }
This calls the method in Logger

This is method in Logger
    public static <T, String>Logger<T> make(T obj, String log) {
        return LoggerImpl.make(obj, log);
    }
which calls

    public static <T, String> Logger<T> make(T obj, String log) {
        if(obj instanceof Logger) {
            throw new IllegalArgumentException("already a Logger");
        } else if (obj == null) {
            throw new IllegalArgumentException("argument cannot be null");
        } 
        return new LoggerImpl<T>(obj, log);
    }

**Comment by Keithlimhs** [comment]
> ```
> public <R> Logger<R> map(Function<? super T, ? extends R> mapper) {
>     printlog();
>     if(mapper.apply(this.obj) == this.obj) {
>         String prompt = String.format("Value unchanged. Value =  %s", this.obj);
>         return make(mapper.apply(this.obj), prompt);
>     }
>     String prompt = String.format("Value changed! New Value = %s", mapper.apply(this.obj));
>     return make(mapper.apply(this.obj), prompt);
> }
> ```
> 
> This calls the method in Logger
> 
> This is method in Logger
> 
> ```
> public static <T, String>Logger<T> make(T obj, String log) {
>     return LoggerImpl.make(obj, log);
> }
> ```
> 
> which calls
> 
> ```
> public static <T, String> Logger<T> make(T obj, String log) {
>     if(obj instanceof Logger) {
>         throw new IllegalArgumentException("already a Logger");
>     } else if (obj == null) {
>         throw new IllegalArgumentException("argument cannot be null");
>     } 
>     return new LoggerImpl<T>(obj, log);
> }
> ```

hmm i didnt overload the make method in Logger, but rather the constructor inside LoggerImpl

**Comment by isaactyj** [comment]
@Keithlimhs but if you dont overload the method in logger how would I take in the log?

**Comment by Keithlimhs** [comment]
> @Keithlimhs but if you dont overload the method in logger how would I take in the log?

I overloaded the constructor for LoggerImpl so i called the overloaded constructor within the map method. The overloaded constructor for LoggerImpl takes in the log

**Comment by ivankqw** [comment]
> @Keithlimhs but if you dont overload the method in logger how would I take in the log?

Yep Keith is right. Should overload the constructor for LoggerImpl.

**Comment by Jason-Ng00** [comment]
Yup Keith is right, so the LoggerImpl will have 2 constructors in which one of them take in only a value while the oither one take in a value and a String log. We can use the former for initialization of a new Logger and the latter one for updating the log

**Comment by isaactyj** [comment]
@ivankqw @Keithlimhs @Jason-Ng00 So in that case would the map function signature be LoggerImpl since you are directly calling it right. How do you store the previous values or logs? Did you use an array or a queue?

**Comment by Keithlimhs** [comment]
> @ivankqw @Keithlimhs @Jason-Ng00 So in that case would the map function signature be LoggerImpl since you are directly calling it right. How do you store the previous values or logs? Did you use an array or a queue?

Hi my map function signature is indeed LoggerImpl since I am calling the LoggerImpl constructor. I store the previous logs as string, which I update during the map function. You think you can use an ArrayList< String > as well to store the previous logs

**Comment by isaactyj** [comment]
@Keithlimhs if you store it in an arraylist, why would your overload constructor not take in an arraylist so you update the arraylist outside before passing it into the constructor? If you pass a single log each time, wouldn't there be a loss of information of the previous few logs?

**Comment by Keithlimhs** [comment]
> @Keithlimhs if you store it in an arraylist, why would your overload constructor not take in an arraylist so you update the arraylist outside before passing it into the constructor? If you pass a single log each time, wouldn't there be a loss of information of the previous few logs?

If you use an ArrayList to store the previous logs, your overloaded method would take in the ArrayList which is updated in the map function

**Comment by Jason-Ng00** [comment]
For me I used an array of String to store the log, and I have a constructor for it. For every time I need to update the log, I will just be adding a String at the back of the array(new).

**Comment by isaactyj** [comment]
@Keithlimhs @Jason-Ng00 Thanks guys for your help!
I used an arraylist but keep getting boolean cannot be converted to arraylist<string> even though there is no boolean

**Comment by Keithlimhs** [comment]
> @Keithlimhs @Jason-Ng00 Thanks guys for your help!
> I used an arraylist but keep getting boolean cannot be converted to arraylist even though there is no boolean

May i see your overloaded constructor?

**Comment by isaactyj** [comment]
This is my map function

    public <R> LoggerImpl<R> map(Function<? super T, ? extends R> mapper) {
        if(mapper.apply(this.obj) == this.obj) {
            String prompt = String.format("Value unchanged. Value =  %s", this.obj);
            this.logs.add(prompt);
            return new LoggerImpl<R>(mapper.apply(this.obj), this.logs);
        }
        String prompt = String.format("Value changed! New value = %s", mapper.apply(this.obj));
        this.logs.add(prompt);
        return new LoggerImpl<R>(mapper.apply(this.obj), logs);
    }

which calls

    protected LoggerImpl(T obj, ArrayList<String> logs) {
        this.obj = obj;
        this.logs = logs;
    }

Im not sure why is it that if map is not a permanent function but my constructor takes it as a permanent function

codecrunch
a.map(x -> x + 1)
$.. ==> Logger[6]
jshell> a.printlog()
Value initialized. Value = 5

mine
a.map(x->x+1)
$.. ==> Logger[6]
a.printlog()
Value initialized. Value = 5
Value changed! new value = 6

**Comment by jethrokyq** [comment]
Hi isaac, you should create a new empty ArrayList within map, copy the logs of the old logger and then update the arraylist with the new entry before passing it to LoggerImpl.

**Comment by Keithlimhs** [comment]
> This is my map function
> 
> ```
> public <R> LoggerImpl<R> map(Function<? super T, ? extends R> mapper) {
>     if(mapper.apply(this.obj) == this.obj) {
>         String prompt = String.format("Value unchanged. Value =  %s", this.obj);
>         this.logs.add(prompt);
>         return new LoggerImpl<R>(mapper.apply(this.obj), this.logs);
>     }
>     String prompt = String.format("Value changed! New value = %s", mapper.apply(this.obj));
>     this.logs.add(prompt);
>     return new LoggerImpl<R>(mapper.apply(this.obj), logs);
> }
> ```
> 
> which calls
> 
> ```
> protected LoggerImpl(T obj, ArrayList<String> logs) {
>     this.obj = obj;
>     this.logs = logs;
> }
> ```
> 
> Im not sure why is it that if map is not a permanent function but my constructor takes it as a permanent function
> 
> codecrunch
> a.map(x -> x + 1)
> $.. ==> Logger[6]
> jshell> a.printlog()
> Value initialized. Value = 5
> 
> mine
> a.map(x->x+1)
> $.. ==> Logger[6]
> a.printlog()
> Value initialized. Value = 5
> Value changed! new value = 6

I think in your overloaded LoggerImpl method, the this.logs should be this.logs = new ArrayList< String > (logs)

**Comment by miao-tianjia** [comment]
not sure if I can ask here, but for the map method, do we simply put it in LoggerImpl or we have to also put it in Logger? I only  put in the LoggerImpl, but I did not pass the test case Logger.make(5).map(x->x+1); but when I try to put the map method in Logger, it says abstract methods should not have body.

**Comment by isaactyj** [comment]
@miao-tianjia it probably mean your map method has a curly bracket?

**Comment by miao-tianjia** [comment]
> @miao-tianjia it probably mean your map method has a curly bracket?

Sorry can you elaborate on this? do you mean that I should have a map method in Logger? I tried to put 

> <U> Logger<U> map(Function<? super T, ? extends U> mapper);
in the Logger interface but it reported the error that I stated above (that abstract method in interface should not have body). if I have to put a map method in Logger, then what should I put? Thank you so much!

**Comment by miao-tianjia** [comment]
@isaactyj I have found out where I was wrong, but thanks a lot anyways!

**Comment by AdelineWongg** [comment]
> > This is my map function
> > ```
> > public <R> LoggerImpl<R> map(Function<? super T, ? extends R> mapper) {
> >     if(mapper.apply(this.obj) == this.obj) {
> >         String prompt = String.format("Value unchanged. Value =  %s", this.obj);
> >         this.logs.add(prompt);
> >         return new LoggerImpl<R>(mapper.apply(this.obj), this.logs);
> >     }
> >     String prompt = String.format("Value changed! New value = %s", mapper.apply(this.obj));
> >     this.logs.add(prompt);
> >     return new LoggerImpl<R>(mapper.apply(this.obj), logs);
> > }
> > ```
> > 
> > 
> > which calls
> > ```
> > protected LoggerImpl(T obj, ArrayList<String> logs) {
> >     this.obj = obj;
> >     this.logs = logs;
> > }
> > ```
> > 
> > 
> > Im not sure why is it that if map is not a permanent function but my constructor takes it as a permanent function
> > codecrunch
> > a.map(x -> x + 1)
> > $.. ==> Logger[6]
> > jshell> a.printlog()
> > Value initialized. Value = 5
> > mine
> > a.map(x->x+1)
> > $.. ==> Logger[6]
> > a.printlog()
> > Value initialized. Value = 5
> > Value changed! new value = 6
> 
> I think in your overloaded LoggerImpl method, the this.logs should be this.logs = new ArrayList(logs)

Hello, may I know why do we have to create a new empty arraylist everytime we update the log? Can we just use the add() method to add the new log at the back?

**Comment by Keithlimhs** [comment]
@AdelineWongg if you do not create a new empty arraylist everytime you update the log, you will be updating the log of the old LoggerImpl instance, which should not be the case because when you call the map, you need to return a new LoggerImpl instance with the updated log while ensuring your previous LoggerImpl instance does not have a change in the logs. One of the test cases provided shows that after the map function is called, the old LoggerImpl instance does not have a change in its log

**Comment by AdelineWongg** [comment]
Oh...I understand now.. Thanks @Keithlimhs  but I get initialisation problem in my 2nd example:
declared :
private final ArrayList<String> logs;

protected LoggerImpl(T t, ArrayList<String> oldLogs) {
    this.t = t;
    this.logs = new ArrayList<String>();
    for (int i =  0; i < oldLogs.size();  i++) {
        logs.add(oldLogs.get(i));
    }
}

I don't understand why variable logs might not have been initialized. I assigned it to an empty ArrayList and updated the contents inside using oldLogs(my input)..

**Comment by Keithlimhs** [comment]
> Oh...I understand now.. Thanks @Keithlimhs but I get initialisation problem in my 2nd example:
> declared :
> private final ArrayList< String > logs;
> 
> ![image](https://user-images.githubusercontent.com/79458766/113821461-b1eaf080-97ae-11eb-8a98-898e0f8fd890.png)
> 
> I don't understand why variable logs might not have been initialized. I assigned it to an empty ArrayList and updated the contents inside using oldLogs(my input)..

I think you have to initialized an ArrayList in the attribute declaration. So it would be **private final ArrayList< String > logs = new ArrayList< String>()**

**Comment by AdelineWongg** [comment]
@Keithlimhs 
After I declare that in my attribute declaration,
in my loggerImpl constructor:
I did this...

![image](https://user-images.githubusercontent.com/79458766/113893887-08315100-97fa-11eb-900a-5cd6ab347c08.png)

but when I a.map(x -> x + 1) 
although it does not update a which is correct, it still updates my original a's  logs which I couldn't understand why because map actually returns a new LoggerImpl
and my printlog simply loop through my this.log to print everything out

**Comment by Keithlimhs** [comment]
@AdelineWongg hi in your constructor you should still initialize a new ArrayList and add the oldList items into the new ArrayList. If i am not wrong you are currently changing the oldList items

**Comment by AdelineWongg** [comment]
@Keithlimhs  but if declared 
private final ArrayList< String > logs = new ArrayList< String >();

In my constructor, I will not be able to do reassignment because logs is final
